README.md

```diff
@@ -151,6 +151,20 @@ A wide range of lockfiles are supported by utilizing this [lockfile package](htt
 $ osv-scanner --lockfile=/path/to/your/package-lock.json --lockfile=/path/to/another/Cargo.lock
 ```
 
+The scanner will attempt to automatically determine the parser to use for each 
+file based on the filename (except for `installed`, see below); you can explicitly
+tell the scanner what parser to use for specific files using the `--parse-as` flag:
+
+```bash
+$ osv-scanner --parse-as 'requirements.txt:/path/to/your/extra-requirements.txt' --lockfile=/path/to/your/extra-requirements.txt
```

Comment:
Should we instead make `--parse-as` just do the parsing itself as well? It seems a little clunky to have to specify the path twice in these cases.

Reply:
It's also rather indirect -- it sets some global state that then needs to be utilised later by specifying the same path again. Not sure if there are benefits to this, and it seems much simpler/easier to use to just streamline this.

Reply:
I'm not too fussed, but note what you're describing sounds based solely on the use of `-L`, when you can pass directories and files as args too.

While still less useful here, as @another-rex pointed out this could be extended to support globs which make this make more sense, i.e. `osv-scanner --parse-as requirements.txt: requirements-*.txt my/project`.

I don't think there's a major loss so like I said I'm not too fussed as I agree it'd be nice to reduce the verbosity, but it might mean there's some edge-cases in future going down this path (I can play around with it tomorrow of course).

A middleground could be to assume `-L` unless ones actually present

Reply:
For the general globbing/directory case, I think that's something that would be better suited as a config file option, rather than trying to support this complex mapping as a CLI option. 

I think we should make `--parse-as` less verbose and just have it do the scanning as well. 

@another-rex thoughts?

Reply:
The problem I see with making parse-as do scanning as well is that it won't be able to support directory scanning at all, which we currently advertise as the easiest way to use osv-scanner.
```
osv-scanner --parse-as requirements.txt:alt_requirement.txt -r my/project
```
will turn into
```
osv-scanner --parse-as requirements.txt:my/project/alt_requirement.txt  --parse-as requirements.txt:my/project/inner/alt_requirement.txt  ...etc.
```
which doesn't look that good.

Reply:
Right, so I see `--parse-as` as a more flexible `-L`, as @G-Rath pointed out. It's an explicit way for someone trying out the tool to use a particular parser on a file they want to test it on. Supporting both `--parse-as` and directory scanning seems a bit messy to me and not the best way we can support this use case. 

For the general directory case, I think the best way to support that is: 
- Improve automated detection by improving our filesystem walking/heuristics. 
- In cases where our automated detection still fails, support this in `osv-scanner`.toml, which seems like a much neater place to do this mapping.

Reply:
That makes sense, do we want to rename `--parse-as` to `--lockfile-parse-as` just to make it clear it's function as a alternative to `--loclfile`?

Reply:
what about just adding this feature then to `-L` i.e. `-L=requirements.txt:/path/to/my/lockfile.txt` and `-L=/path/to/my/requirements.txt`.

That would then free up `--parse-as` to still stick around for similar glob based usage with directory - arguably it could be a breaking change because directories with `:` in them would trigger this, but it feels like that would be worth it especially when we could mitigate this by saying `-L=:/path/to/my/lockfile.txt` means "use default/inferred parser" and then folks just need to do that to handle that case (+ I wouldn't expect that to be very common in directory/file names).

Reply:
That was one of the original proposals in https://github.com/google/osv-scanner/pull/94#discussion_r1060334625 :) 

My concern there was that it was hard to parse perfectly in all cases. e.g. what if we had ':' characters in the given pathnames? Or if we had a filename called "requirements.txt:foo" ? Unlikely, but it does feel like a gap.

Reply:
🤦 I completely missed that somehow.

I think it would be fine, and I provided a way to handle edgecases: if we say that the split is always on the first `:` and that an empty "parse-as" means "use default/infer", then that should prevent any edgecases - even if your file started with an `:`, you'd just provide `-L=::my-file`.

If that doesn't make sense, but you like the idea I can switch and then you can see it in actual action via the tests

Reply:
Hmm, I'm still not sure that handles the edge case where the full filename is "requirements.txt:foo". 

In this case, perhaps we can expand this into multiple files if needed. i.e.

- if "requirements.txt:foo" exists, scan that through the default/infer rules. 
- if "foo" also exists, scan "foo" as "requirements.txt" as well.

Reply:
It should because you'd be providing `-L=:requirements.txt:foo`

Reply:
Right, but I think that would be a slight behaviour change, because previously scanning "requirements.txt:foo" would scan that file directly. 

That said, it's enough of an edge case that I don't think anyone would actually be affected by this, so what you suggested seems good to me.